pkg/prom/instance/modal_manager.go

```diff
@@ -0,0 +1,182 @@
+package instance
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// Mode controls how instances are created.
+type Mode string
+
+// Types of instance modes
+var (
+	ModeDistinct Mode = "distinct"
+	ModeShared   Mode = "shared"
+
+	DefaultMode = ModeShared
+)
+
+// UnmarshalYAML unmarshals a string to a Mode. Fails if the string is
+// unrecognized.
+func (m *Mode) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*m = DefaultMode
+
+	var plain string
+	if err := unmarshal(&plain); err != nil {
+		return err
+	}
+
+	switch plain {
+	case string(ModeDistinct):
+		*m = ModeDistinct
+		return nil
+	case string(ModeShared):
+		*m = ModeShared
+		return nil
+	default:
+		return fmt.Errorf("unsupported instance_mode '%s'. supported values 'shared', 'distinct'", plain)
+	}
+}
+
+// ModalManager runs instances by either grouping them or running them fully
+// separately.
+type ModalManager struct {
+	mut     sync.RWMutex
+	mode    Mode
+	configs map[string]Config
+
+	currentActiveConfigs prometheus.Gauge
+
+	log log.Logger
+
+	// Next is the underlying manager this manager wraps.
+	next        Manager
+	modeManager Manager
+}
+
+// NewModalManager creates a new ModalManager.
+func NewModalManager(reg prometheus.Registerer, l log.Logger, next Manager, mode Mode) (*ModalManager, error) {
+	if mode == "" {
+		mode = DefaultMode
+	}
+
+	currentActiveConfigs := promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name: "agent_prometheus_active_configs",
+		Help: "Current number of active configs being used by the agent.",
+	})
+
+	mm := ModalManager{
+		next:                 next,
+		log:                  l,
+		currentActiveConfigs: currentActiveConfigs,
+		configs:              make(map[string]Config),
+	}
+	if err := mm.SetMode(mode); err != nil {
+		return nil, err
+	}
+	return &mm, nil
+}
+
+// SetMode updates the mode ModalManager is running in. Changing the mode is
+// an expensive operation; all underlying configs must be stopped and then
+// reapplied.
+func (m *ModalManager) SetMode(newMode Mode) error {
+	m.mut.Lock()
+	defer m.mut.Unlock()
+
+	if m.mode == newMode {
+		return nil
+	}
+	m.mode = newMode
+
+	// Stop the old mode before changing it. It won't exist if this is the first
+	// time calling SetMode from NewModalManager.
+	if m.modeManager != nil {
+		m.modeManager.Stop()
+	}
+
+	switch m.mode {
```

Comment:
This isn't really a concrete suggestion, but: `mode`, `next`, and `modeManager` seem to be their own distinct kind of "state" for the `ModalManager` that always need to be coordinated together. Is there a way they could be extracted to another object that would prevent misuse or make `ModalManager` a bit easier to reason about?

Reply:
Hmm, not sure I'm following. Do you have any examples?

Reply:
Not exactly, no. My feeling though is that these three fields make up a state machine and so it might make sense to make that explicit.

Maybe something like 

```go
type ModalState struct {
    mode Mode
    wrapped Manager
    active Manager
 }
 
 func (s *ModalState) ToDistinct() error {
     if s.mode == DistinctMode {
         return nil
     }
     
    // transition logic modifying `wrapped` and `active`
}

func (s *ModalState) ToShared() error {
    if s.mode == SharedMode {
        return nil
    }
    
    // transition logic modifying `wrapped` and `active`
}

func (s *ModalState) Manager() Manager {
    return s.active
} 

func (m *ModalManager) SetMode(mode Mode) error {
    m.mtx.Lock()
    defer m.mtx.Unlock()
    
    if mode == DistinctMode {
        s.ToDistinct()
    } else if (mode == SharedMode) {
        s.ToShared()
    } else {
        // panic, etc.
    }
    
    active := s.Manager()
    for name, cfg := range m.configs {
        active.ApplyConfig(cfg) // ... etc
    }
    
    // etc
}
```

Reply:
Hmm, let me sit on this for a bit. I want to eventually deprecate the instance sharing (#398) so I don't want to spend too much time on trying to make this clean if it'll be removed.

Reply:
(In general, I'm not thrilled about this instance Manager chaining and I'm eager to throw it out completely with something simpler)